sql/rpl_replica.cc

```diff
@@ -6099,6 +6099,20 @@ static void *handle_slave_worker(void *arg) {
                                                     // initialization with
                                                     // `PRIVILEGE_CHECKS_USER`
 
+#ifdef WITH_WSREP
+  /* Initialization of Galera/WSREP and async replication happens in parallel.
+     Allow async replication infrastructure to init up to this point, but
+     wait with applying async-replicated events until WSREP infrastructure
+     is fully initialized (wsrep_ready == true). wsrep_ready is set when
+     server state shifts to s_synced.
+     Before this state, events are rejected with error
+     'WSREP has not yet prepared node for application use' */
+  if (WSREP(thd)) {
+    Wsrep_server_state::instance().wait_until_state(
+        Wsrep_server_state::s_synced);
+  }
```

Comment:
This code gets executed only when the server is configured with `--replica-parallel-workers>1` i.e, in  multi-threaded replica mode. If there is only one async applier thread running on replica, then we may still get the issue. I think moving this wait to `handle_slave_sql()` before we call `slave_start_workers()` would handle both single-threaded and multi-threaded replicas.

Reply:
Right! Fixed.